linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -381,3 +387,61 @@ static int threadpool_work_func(void* param);
+```
+
+`threadpool_create_work_item` creates a work item to be executed by the threadpool.
+
+**S_R_S THREADPOOL_LINUX_05_001: [** If `threadpool` is `NULL`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_002: [** If `work_function` is `NULL`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_003: [** `threadpool_create_work_item` shall call `sm_exec_begin`. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_004: [** If `sm_exec_begin` returns `SM_EXEC_REFUSED`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_005: [** `threadpool_create_work_item` shall acquire the `SRW lock` in shared mode by calling `srw_lock_acquire_shared`. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_006: [** `threadpool_create_work_item` shall increment the `insert_pos`. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_007: [** If task state is `TASK_NOT_USED`, `threadpool_create_work_item` shall set the current task state to `TASK_INITIALIZING`. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_008: [** Otherwise, `threadpool_create_work_item` shall release the shared SRW lock by calling `srw_lock_release_shared` and increase `task_array` capacity: **]**
+
+**S_R_S_THREADPOOL_LINUX_05_009: [** If reallocating the task array fails, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_010: [** `threadpool_create_work_item` shall copy the work function and work function context into insert position in the task array and return the variable of `PTHREADPOOL_TASK` at insert position. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_011: [** `threadpool_create_work_item` shall set the `task_state` to `TASK_WAITING` and then release the shared SRW lock. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_012: [** `threadpool_create_work_item` shall return the `PTHREADPOOL_TASK` at insert position. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_013: [** `threadpool_create_work_item` shall call `sm_exec_end`. **]**
+
+```c
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_ITEM_HANDLE, work_item_context);
+```
+
+`threadpool_schedule_work_item` schedules a work item to be executed by the threadpool.
+
+**S_R_S_THREADPOOL_LINUX_05_014: [** `work_item_context` can be a `NULL` value **]**
+
+**S_R_S_THREADPOOL_LINUX_05_015: [** If `threadpool` is `NULL`, `threadpool_schedule_work_item` shall fail and return a non-zero value. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_016: [** `threadpool_schedule_work_item` shall call `sm_exec_begin`. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_017: [** If `sm_exec_begin` returns `SM_EXEC_REFUSED`, `threadpool_schedule_work_item` shall fail and return a non-zero value. **]**
+
+**S_R_S_THREADPOOL_LINUX_05_018: [** `threadpool_schedule_work_item` shall unblock the `threadpool` semaphore by calling `sem_post`. **]**
```

Comment:
>sem_post [](http://example.com/codeflow?start=122&length=8)

still need to figure out how we will be able to call this multiple times and have the work execute each time.

Reply:
Linux implementation would be safe because all we are doing is creating additional work items and calling schedule. So each call to schedule should have done items to work on

Reply:
We can have a chat about this offline so I understand the issue.